lib/grack/git.rb

```diff
@@ -0,0 +1,65 @@
+module Grack
+  class Git
+
+    attr_reader :repo
+
+    def initialize(git_path, repo_path)
+      @git_path = git_path
+      @repo = repo_path
+    end
+
+    def update_server_info
+      execute(%W(update-server-info))
+    end
+
+    def command(cmd)
+      [@git_path || 'git'] + cmd
+    end
+
+    def capture(cmd)
+      IO.popen(popen_env, cmd, popen_options).read
+    end
+
+    def execute(cmd)
+      if block_given?
+        IO.popen(popen_env, command(cmd), File::RDWR, popen_options) do |pipe|
+          yield(pipe)
+        end
+      else
+        capture( command(cmd) ).chomp
+      end
+    end
+
+    def popen_options
+      {chdir: repo, unsetenv_others: true}
+    end
+
+    def popen_env
+      {'PATH' => ENV['PATH'], 'GL_ID' => ENV['GL_ID']}
+    end
+
+    def config_setting(service_name)
+      service_name = service_name.gsub('-', '')
+      setting = config("http.#{service_name}")
+      if service_name == 'uploadpack'
+        return setting != 'false'
+      else
+        return setting == 'true'
+      end
+    end
+
+    def config(config_name)
+      execute(%W(config #{config_name}))
+    end
+
+    def valid_repo?
+      return false unless (File.exists?(repo) && File.realpath(repo) == repo)
+      match = execute(%W(rev-parse --git-dir)).match(/\.$|\.git$/)
+      if match.to_s == '.git'
+        # Since the parent could be a git repo, we want to make sure the actual repo contains a git dir.
+        return false unless Dir.entries(repo).include?('.git')
+      end
+      return !!match
```

Comment:
@DouweM Fair enough, thats just a style thing I've adopted ( not sure how good of a practice it is ) to make boolean methods that return at different points more readable and obvious. 

And I'll get to work on revamping the tests today. Any reason why I shouldn't convert to Rspec to match the rest of Gitlab? or do you want to stay more inline the original Grack?